x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenizer.java

```diff
@@ -0,0 +1,447 @@
+import java.util.stream.IntStream;
+
+public class BpeTokenizer extends Tokenizer {
+    private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
+    private final OffsetAttribute offsetAtt = addAttribute(OffsetAttribute.class);
+    private final PositionIncrementAttribute posIncAtt = addAttribute(PositionIncrementAttribute.class);
+
+    private static final Pattern pattern = Pattern.compile(
+        "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)|\\s+"
+    );
+
+    static Map<Integer, Character> byteEncoder() {
+        List<Integer> bytes = IntStream.concat(
+            IntStream.range(Character.codePointAt("!", 0), Character.codePointAt("~", 0) + 1),
+            IntStream.concat(
+                IntStream.range(Character.codePointAt("¡", 0), Character.codePointAt("¬", 0) + 1),
+                IntStream.range(Character.codePointAt("®", 0), Character.codePointAt("ÿ", 0) + 1)
+            )
+        ).boxed().collect(Collectors.toList());
+        List<Integer> chars = new ArrayList<>(bytes);
+        int n = 0;
+        for (int i = 0; i < 2 << 7; i++) {
+            if (bytes.contains(i)) {
+                continue;
+            }
+            bytes.add(i);
+            chars.add((2 << 7) + n);
+            n++;
+        }
+        char[] charArray = new char[chars.size()];
+        int i = 0;
+        for (Integer c : chars) {
+            charArray[i++] = Character.toChars(c)[0];
+        }
+        Map<Integer, Character> map = new HashMap<>(charArray.length, 1.0f);
+        for (int j = 0; j < bytes.size(); j++) {
+            map.put(bytes.get(j), charArray[j]);
+        }
+        return map;
+    }
+
+    private static final Map<Integer, Character> BYTES_CHAR = byteEncoder();
+
+    // NOTE: 32 is " " utf-8 encoded byte
+    private static final char ENCODED_SPACE_CHAR = BYTES_CHAR.get(32);
+
+    public static BpeTokenizer build(
+        List<String> neverSplit,
+        List<String> dictionary,
+        List<String> merges,
+        String unknownToken,
+        boolean isPrefixSpace
+    ) {
+        CharArraySet neverSplitSet = new CharArraySet(neverSplit, false);
+        CharTrie neverSplitTree = CharTrie.build(neverSplit);
+        CharArrayMap<Integer> mergeRanks = new CharArrayMap<>(merges.size(), false);
+        int mergePos = 0;
+        for (String merge : merges) {
+            mergeRanks.put(Strings.replace(merge, " ", ""), mergePos++);
+        }
+        CharArrayMap<Integer> vocabHash = new CharArrayMap<>(dictionary.size(), false);
+        int vocabPos = 0;
+        for (String v : dictionary) {
+            vocabHash.put(v, vocabPos++);
+        }
+        return new BpeTokenizer(isPrefixSpace, mergeRanks, neverSplitSet, neverSplitTree, vocabHash, unknownToken);
+    }
+
+    private final StringBuilder str = new StringBuilder();
+    // This is pretty much stolen from PatternTokenizer. We really should stream the input and stop using regex
```

Comment:
Yes, I think it's worth doing this in a followup PR before moving on to the next project.

The regex is pretty simple. It basically boils down to one of:

1. An ASCII apostrophe followed by one of a handful of specific letter sequences
2. Optional space followed by an unbroken sequence of letters
3. Optional space followed by an unbroken sequence of numbers
4. Optional space followed by an unbroken sequence of characters that are neither letters, numbers nor whitespace
5. Non-empty sequence of whitespace

There would be large benefits in being able to stream the input instead of making multiple copies, and a loop that looks for one of those 5 options should also be quite simple.